### PR TITLE
Selected items list

### LINKED ISSUE #3118: Feature Request: SelectedItemsList Component
Create a component that only takes cares of selected items. Breaking responsibilities of the BasePicker down further.

In conjunction with the FloatingPicker #3043 we are able to create a new component, like the BasePicker, but more compartmentalized.

Priorities and help requested (not applicable if asking question):
Are you willing to submit a PR to fix? Yes!

### PR DESCRIPTION
**Pull request by amyngu**
#### Pull request checklist

- [x] Addresses an existing issue: Fixes #3118 
- [x] Include a change request file using `$ npm run change`

#### Description of changes
- Modified folder names (ExtendedPicker -> FloatingPicker)
- Created `SelectedItemsList`, with copy support
- Changed `BaseExtendedPicker` from extending from `BasePicker` (in office-ui-fabric-react) to being its own component utilizing `SelectedItemsList` and `BaseFloatingPicker`
- Created `SelectedPeopleList` (extends `SelectedItemsList`) support expand group behavior
- Added `ExtendedSelectedItem` for expanding group personas and right click menu support with copy and remove options

### CODE REVIEW

**Comment by micahgodbolt** [comment]
@amyngu these are some pretty extensive changes. Can you give us a little background on where these changes came from? Was this a product driven change, and if so, which product, and what need?

**Comment by amyngu** [comment]
@micahgodbolt in OWA we wanted a more robust PeoplePicker that fits our needs that we think others can benefit from. We also wanted to fabric control that was just the floating picker (for purposes of @mentions). I discussed this through with @joschect and he suggested breaking up the responsibilities of what currently is the BasePicker and create two components, SelectedItemsList and a FloatingPicker (done in a previous PR) and use those two to create a new picker, not touching any logic from the current BasePicker. I hope this clears up some of your questions. Please feel free to let me know if you still have open questions.

**Comment by micahgodbolt** [comment]
@amyngu that's great. just good to hear what discussions lead to this change, and that more than 1 person was involved in the decisions. Can you take a look at the current build error?

**Comment by amyngu** [comment]
@ThomasMichon & @joschect I'd appreciate some reviews on this. Thanks!

File: packages/experiments/src/components/SelectedItemsList/SelectedPeopleList/Items/ExtendedSelectedItem.tsx [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+            className={ css('ms-PickerItem-removeButton', styles.expandButton, styles.actionButton) }
+            ariaLabel={ removeButtonAriaLabel }
+          />
+        </div>
+        <div className={ css(styles.personaWrapper) }>
+          <div
+            className={ css('ms-PickerItem-content', styles.itemContent) }
+            id={ 'selectedItemPersona-' + itemId }
+          >
+            <Persona
```
Comment (joschect) [inline packages/experiments/src/components/SelectedItemsList/SelectedPeopleList/Items/ExtendedSelectedItem.tsx]: Is this supposed to be for any item picker? Or only for people pickers? If it's generic then this shouldn't be a persona, it should be a renderer that can get passed in.

File: packages/experiments/src/components/SelectedItemsList/SelectedPeopleList/Items/ExtendedSelectedItem.tsx [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+export class ExtendedSelectedItem extends BaseComponent<ISelectedPeopleItemProps, IPeoplePickerItemState> {
+  protected persona: HTMLElement;
+
+  private menuItems: IContextualMenuItem[];
+
+  constructor(props: ISelectedPeopleItemProps) {
+    super(props);
+    this.state = { contextualMenuVisible: false };
+    let { onCopyItem, onRemoveItem, copyMenuItemText, removeMenuItemText } = this.props;
+    this.menuItems = [
```
Comment (joschect) [inline packages/experiments/src/components/SelectedItemsList/SelectedPeopleList/Items/ExtendedSelectedItem.tsx]: Why not let the consumer pass in other menu items?

Comment (joschect) [reply packages/experiments/src/components/SelectedItemsList/SelectedPeopleList/Items/ExtendedSelectedItem.tsx]: I'm actually pretty sure someone will want to pass in their own menu items.

File: packages/experiments/src/components/SelectedItemsList/BaseSelectedItemsList.tsx [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+  /**
+   * Controls what happens whenever there is an action that impacts the selected items.
+   * If selectedItems is provided as a property then this will act as a controlled component and it will not update it's own state.
+  */
+  protected async updateSelectedItems(items: T[], focusIndex?: number): Promise<void> {
+    if (this.props.selectedItems) {
+      // If the component is a controlled component then the controlling component will need
+      this.onChange(items);
+    } else {
+      await this.setState({ items: items }, () => {
```
Comment (joschect) [inline packages/experiments/src/components/SelectedItemsList/BaseSelectedItemsList.tsx]: Why is this async and why are you using await? There are no instances where you do updateSelectedItems().then

File: packages/experiments/src/components/SelectedItemsList/BaseSelectedItemsList.tsx [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+  }
+
+  protected onSelectionChange(): void {
+    this.forceUpdate();
+  }
+
+  protected onChange(items?: T[]): void {
+    if (this.props.onChange) {
+      // tslint:disable-next-line:no-any
+      (this.props.onChange as any)(items);
```
[the remark this answers is not in the document]
Comment (amyngu) [reply packages/experiments/src/components/SelectedItemsList/BaseSelectedItemsList.tsx]: Will change the actual signature of the property

File: packages/experiments/src/components/SelectedItemsList/BaseSelectedItemsList.tsx (hunk 3 above)
Comment (amyngu) [reply packages/experiments/src/components/SelectedItemsList/BaseSelectedItemsList.tsx]: Will remove async and await

File: packages/experiments/src/components/SelectedItemsList/SelectedPeopleList/Items/ExtendedSelectedItem.tsx (hunk 1 above)
Comment (amyngu) [reply packages/experiments/src/components/SelectedItemsList/SelectedPeopleList/Items/ExtendedSelectedItem.tsx]: This is only for a people section scenario (file is located under SelectdPeopleList/Items).

File: packages/experiments/src/components/SelectedItemsList/SelectedPeopleList/Items/ExtendedSelectedItem.tsx (hunk 2 above)
Comment (amyngu) [reply packages/experiments/src/components/SelectedItemsList/SelectedPeopleList/Items/ExtendedSelectedItem.tsx]: You're right. I exposed the menuItems as a prop, so they can be passed in

**Comment by amyngu** [comment]
@dzearing Can I get this merged?

**Comment by dzearing** [comment]
Picker stuff should have @joschect involved

**Pull request merged**